src/editor.rs

```diff
@@ -6,16 +6,15 @@ use crate::context::ColorClosure;
 use crate::cursor::CursorPosition;
 use crate::event::*;
 use crate::prompt::Prompt;
-use crate::{util, Terminal};
+use crate::{util, Completer, EditorRules, Terminal};
 use crate::{Buffer, Cursor};
 use crate::{History, Metrics};
 
-use super::complete::Completer;
-
 /// The core line editor. Displays and provides editing for history and the new buffer.
 pub struct Editor<'a> {
     prompt: Prompt,
     history: &'a mut History,
+    editor_rules: &'a dyn EditorRules,
```

Comment:
You could use a generic here and avoid the lifetime.  I.e. pub struct Editor<T: EditorRules> ... editor_rules: T and hand ownership to Editor.  Not strictly needed but make it a little easier to use Editor if you don't need worry about the lifetime tied to an EditorRules instance.

Reply:
ok, so, initially I avoided this because introducing a type parameter to editor was a silly amount of signature changes. and my main goal was to make sure the DefaultEditorRule was using generics such that it was easier for the client to move their own implementations in:
```rust 
    let mut con = Context::new();
    let editor_rules = DefaultEditorRules::custom(
        DefaultWordDivideRule {},
        NewlineForBackslashAndOpenDelimRule {},
    ); // this is the part I used generics for, and I like that the client just move their implementations in
    con.set_editor_rules(Box::new(editor_rules)); // but we're still using a trait object type here
```

Today I bit the bullet and started changing everything in Editor to be generic and ended up in a weird corner. I put the code in  a separate branch because we eventually might want this and it's very tedious to do: 

https://github.com/sl-sh-dev/sl-liner/tree/feature/enhance-multiline-genericize

the compilation is failing because Completer and KeyMap are not type object safe when they have methods that use generics. I think I understand why: 

https://doc.rust-lang.org/reference/items/traits.html#object-safety

but to actually make Editor use a generic EditorRules, assuming KeyMap and Completer are going to need to refer to the Editor, which is 100% going to be true going forward, then KeyMap and Completer are also going to have to be made generic (right?), and that might be out of scope for this PR. And I'm not sure the sheer amount of code changes justify the win of not having a lifetime.